File: memoization/golang/01_fib.go
// Fibonacci.

// Implementation of the fibonacci function using memoization.

// The fibonacci function takes in a number as an argument.
// The function should return the n-th number of the Fibonacci sequence.
// The first and second number of the sequence is 1.
package main

import "fmt"

// Find the n-th number of the Fibonacci sequence.
func Fibonacci(n int, memo *map[int]int) int {
	if ans, ok := (*memo)[n]; ok {
		return ans
	}

	if n <= 2 {
		return 1
	}

	(*memo)[n] = Fibonacci(n-1, memo) + Fibonacci(n-2, memo)
	return (*memo)[n]
}

func main() {
	memo := make(map[int]int)

	fmt.Println(Fibonacci(6, &memo))  // 8
	fmt.Println(Fibonacci(7, &memo))  // 13
	fmt.Println(Fibonacci(8, &memo))  // 21
	fmt.Println(Fibonacci(50, &memo)) // 12586269025
}

File: memoization/python/01_fib.py
"""Fibonacci.

This script implements a fibonacci function using memoization.

The fibonacci function takes in a number as an argument. The function should return the
n-th number of the Fibonacci sequence. The first and second number of the sequence is 1.
"""


def fibonacci(n: int, memo: dict = None) -> int:
    """Given a integer n, return the n-th number of the fibonacci sequence.

    Args:
        n (int): The index of a number in the fibonacci sequence.
        memo (dict): A hashmap to cache combinations already calculated.

    Returns:
        int: The n-th number of the fibonacci sequence.
    """
    if memo is None:
        memo = {}

    if n in memo:
        return memo[n]

    if n < 2:
        return n

    memo[n] = fibonacci(n - 1, memo) + fibonacci(n - 2, memo)

    return memo[n]


if __name__ == "__main__":
    print(fibonacci(6))  # 8
    print(fibonacci(7))  # 13
    print(fibonacci(8))  # 21
    print(fibonacci(50))  # 12586269025
